plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/DiagramToXtextCommands.java

```diff
@@ -195,135 +200,152 @@ private static void generateSequenceMessages(Model domainModel, Scenario scenari
-          }          
+          }
         }
       } else if (ends[i] instanceof ExecutionEnd) {
         EObject participantDeactivateMsg = getParticipantDeactivationMsgFromExecutionEnd(ends[i], factory);
-        
+
         // add the deactivation, to the model, or to a block
-        if(blockConditons.isEmpty()) {
+        if (blockConditons.isEmpty()) {
           messagesOrReferences.add(participantDeactivateMsg);
-        }
-        else {
+        } else {
           blockConditons.peek().getBlockElements().add((Message) participantDeactivateMsg);
         }
-        
+
         updateMessagesToDeactivate(messagesToDeactivate);
         i = i + 1;
       } else if (ends[i] instanceof FragmentEnd) {
-        if(ends.length > i+1 && ends[i+1] instanceof InteractionOperand) {
+        if (ends.length > i + 1 && ends[i + 1] instanceof InteractionOperand) {
           // when we encounter a combination of FragmentEnd + Interaction operand, generate alt sequence
-          Alt alt = createAlt(factory, (InteractionOperand)ends[i+1]);
+          Alt alt = createAlt(factory, (InteractionOperand) ends[i + 1]);
           conditions.push(alt);
-          
+
           // add the new encountered alt, to the model, or to a block
-          if(blockConditons.isEmpty()) {
+          if (blockConditons.isEmpty()) {
             domainModel.getElements().add(alt);
-          }
-          else {
+          } else {
             blockConditons.peek().getBlockElements().add(alt);
           }
-          
+
           Block altBlock = createBlock(factory);
           blockConditons.push(altBlock);
           alt.setBlock(altBlock);
-         
+
           i++;
-        }
-        else {
+        } else {
           // here is the end of the alt sequence, extract the last processed alt and its last block
-          if(!conditions.empty()) conditions.pop();
-          if(!blockConditons.empty()) blockConditons.pop();
+          if (!conditions.empty())
+            conditions.pop();
+          if (!blockConditons.empty())
+            blockConditons.pop();
         }
         i++;
-      } else if(ends[i] instanceof InteractionOperand) {
+      } else if (ends[i] instanceof InteractionOperand) {
         // the previous operation block is ended, extract it from the stack, we are done with it
-        if(!blockConditons.empty()) blockConditons.pop();
-        
+        if (!blockConditons.empty())
+          blockConditons.pop();
+
         // generate a new branch for alt (else sequence)
-        Block altBlock = addAltBlock(factory, conditions.peek(), (InteractionOperand)ends[i]);
-        
+        Block altBlock = addAltBlock(factory, conditions.peek(), (InteractionOperand) ends[i]);
+
         blockConditons.push(altBlock);
-        
+
         i++;
-      }
-      else {
+      } else if (ends[i] instanceof InteractionState) {
+        for (EObject timeLapse : timeLapses) {
+          if (timeLapse instanceof StateFragmentImpl) {
```

Comment:
Work with the interface if possible instead of the Impl, use StateFragment insetad of StateFragmentImpl.